### ISSUE TITLE
Godot doesn't support custom screen resolution

### ISSUE THREAD

**Issue by miskatonicstudio** [godotengine/godot#24720]
**Godot version:**

3.0.6

**OS/device including version:**

Linux Pop!_OS 17.10 (Artful Aardvark)

**Issue description:**

For now it seems that Godot supports 2 modes for 3D: fixed resolution (with fullscreen off) or stretch (covering the entire screen, with fullscreen on). Modern games often offer a specific resolution in fullscreen (being able to play with 1280x720 when the screen resolution is 1920x1080). Is this supported in Godot too?

**Steps to reproduce:**

**Minimal reproduction project:**

**Comment by Calinou** [comment]
> Is this supported in Godot too?

Yes, using the `viewport` stretch mode; see [Multiple resolutions](https://docs.godotengine.org/en/latest/tutorials/viewports/multiple_resolutions.html) in the documentation.

Note that this will use nearest-neighbor upscaling, which isn't suited to games that don't follow a "pixel art" aesthetic. It's possible to create a custom viewport to use linear filtering instead, but the Project Settings currently don't offer a way to do that out of the box.

**Comment by miskatonicstudio** [comment]
I'm afraid that's not what I meant. The game should be able to run in a custom resolution in fullscreen, as long as that resolution is supported by the monitor. Example:
* open the game in fullscreen in full HD (1920x1080), which is the maximum resolution of the monitor
* take a screenshot => it saves as an image 1920x1080
* still in fullscreen, change the game's resolution to 1280x720 (assuming that the monitor supports it)
* take a screenshot => it saves as an image 1280x720

The problem is not with stretching the content of the screen, but with actually changing the resolution of the screen. Is such a thing possible with Godot?

**Comment by Calinou** [comment]
Non-native resolutions would require support for exclusive fullscreen, which is not currently supported by Godot: https://github.com/godotengine/godot/issues/14542

Keep in mind there are many practical downsides to using exclusive fullscreen, so modern games tend to be moving away from it.

**Comment by miskatonicstudio** [comment]
It's been a while, but while working on a new project I've noticed how much the game slows down on my very-not-high-end machine. Has there been any development/any decisions made with implementing exclusive fullscreen in Godot?

**Comment by Calinou** [comment]
@miskatonicstudio No, there are still no plans to add exclusive fullscreen to Godot. You can change your display resolution manually if you run into performance issues. It's not as convenient, but considering it's an edge case, I don't see much point in making it a built-in feature (similar to features like hardware gamma).

**Comment by Calinou** [comment]
I just noticed this issue is a duplicate of #14542, closing.

**Comment by miskatonicstudio** [comment]
Thank you for your answer @Calinou . I don't mind changing the screen's resolution before running a game (I'm a Linux user, I'm used to all kinds of workarounds ;) ), but I don't think I can demand that of users/gamers. I think most games nowadays still support changing the resolution of the screen in fullscreen mode, to make your game run faster or the interface elements look bigger. I can understand, however, that Godot doesn't want to add this feature. I will experiment with viewport scaling and see if I can get the performance I want ;)

**Comment by Calinou** [comment]
> to make [...] the interface elements look bigger. 

This is handled automatically by Godot if you use the `2d` or `viewport` stretch mode. See [Multiple resolutions](https://docs.godotengine.org/en/latest/tutorials/viewports/multiple_resolutions.html) for more information. If you'd like to control the scaling factor manually when using the `2d` stretch mode, this will be implemented once https://github.com/godotengine/godot/pull/21446 is merged.
